cpp/inc/bond/protocol/detail/protobuf_utils.h

```diff
@@ -0,0 +1,299 @@
+        }
+
+        if (std::strcmp(it->second.c_str(), "Fixed") == 0)
+        {
+            return Encoding::Fixed;
+        }
+
+        if (std::strcmp(it->second.c_str(), "ZigZag") == 0)
+        {
+            return Encoding::ZigZag;
+        }
+
+        InvalidEnumValueException(name.c_str(), "Encoding");
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata, const std::string& name)
+    {
+        switch (type)
+        {
+        case BT_FLOAT:
+        case BT_DOUBLE:
+            return Encoding::Fixed;
+
+        case BT_UINT8:
+        case BT_UINT16:
+        case BT_UINT32:
+        case BT_UINT64:
+        case BT_INT8:
+        case BT_INT16:
+        case BT_INT32:
+        case BT_INT64:
+            return ReadEncoding(name, metadata);
+
+        case BT_WSTRING:
+        case BT_SET:
+        case BT_MAP:
+            NotSupportedException("Wide string or nested set/map");
+
+        case BT_BOOL:
+        case BT_LIST:
+        case BT_STRING:
+        case BT_STRUCT:
+        default:
+            return Unavailable<Encoding>();
+        }
+    }
+
+    inline Encoding ReadEncoding(BondDataType type, const Metadata& metadata)
+    {
+        return ReadEncoding(type, metadata, AttributeName<>::encode);
+    }
+
+    inline Encoding ReadKeyEncoding(BondDataType type, const Metadata& metadata)
+    {
+        if (type == BT_FLOAT || type == BT_DOUBLE)
+        {
+            NotSupportedException("Float/double map key");
+        }
+
+        return ReadEncoding(type, metadata, AttributeName<>::encode_key);
+    }
+
+    inline Encoding ReadValueEncoding(BondDataType type, const Metadata& metadata)
+    {
+        return ReadEncoding(type, metadata, AttributeName<>::encode_value);
+    }
+
+    inline Packing ReadPacking(const std::string& name, const Metadata& metadata)
+    {
+        auto it = metadata.attributes.find(name);
+        if (it == metadata.attributes.end())
+        {
+            return Unavailable<Packing>();
+        }
+
+        if (std::strcmp(it->second.c_str(), "False") == 0)
+        {
+            return Packing::False;
+        }
+
+        InvalidEnumValueException(name.c_str(), "Packing");
+    }
+
+    inline Packing ReadPacking(BondDataType type, const Metadata& metadata)
+    {
+        switch (type)
+        {
+        case BT_STRING:
+        case BT_STRUCT:
+        case BT_LIST:
+            return Packing::False;
+
+        case BT_WSTRING:
+        case BT_SET:
+        case BT_MAP:
+            NotSupportedException("Wide string or nested set/map");
+
+        default:
+            return ReadPacking(AttributeName<>::pack, metadata);
+        }
+    }
+
+    inline uint32_t MakeTag(uint16_t id, WireType wireType)
+    {
+        BOOST_ASSERT(wireType != Unavailable<WireType>());
+
+        if (id == 0 || (id >= 19000 && id <= 19999))
+        {
+            NotSupportedException("Field ordinal with value 0 or in the range 19000-19999");
```

Comment:
It's worth leaving a comment for us that protobuf fields count from 1. What's the deal with 19000-19999?

Reply:
A reserved range: https://developers.google.com/protocol-buffers/docs/proto3#assigning-tags
> The smallest tag number you can specify is 1, and the largest is 229 - 1, or 536,870,911. You also cannot use the numbers 19000 through 19999

Reply:
All NotSupportedExceptions mean that ProtoBuf does not support it? Could it be confusing since it could be the case that Bond does not support it currently or that Protobuf does not support it? Meaning that we would need something that explains further, a better exception message or a comment explaining why, etc.

Reply:
This is a good idea. Let's draw a distinction between "we haven't done it yet" and "we don't think there's a good way to do it".

Reply:
From writer's perspective it is not supported by protobuf (except of `oneof` which is obvious). Please suggest a better phrasing for exception/message to make it clear.

Reply:
What do you think about `NotSupportedException` vs. `NotImplementedException`?

To be clear, I don't think we need the second one if the final version of this change only has exceptions of the first time. The wstring support is the only thing that jumps out at me as belonging to the second type.

Reply:
Well, the `wstring` is more in the same group with nested containers for me (supported on bond, not on protobuf), but I'll be implementing a conversion for it as you suggested. that sounds reasonable.